File: frontend/src/lib/wireTypes.js
// Mirror of api/app/wire_types.py. The backend is the authority — this exists only to
// fail before spending a network round-trip and a sandbox run. Keep the two in step.
//
// Grammar: T := int | float | str | bool | None | Any | List[T] | Dict[str, T] | Optional[T] | T | None

const SCALARS = new Set(["int", "float", "str", "bool", "none", "any"]);
const BARE_CONTAINERS = new Set(["list", "dict"]);

function splitTopLevel(text, sep) {
  const parts = [];
  let depth = 0;
  let current = "";
  for (const ch of text) {
    if (ch === "[") depth++;
    else if (ch === "]") depth--;
    if (ch === sep && depth === 0) {
      parts.push(current);
      current = "";
    } else {
      current += ch;
    }
  }
  parts.push(current);
  return parts.map((p) => p.trim());
}

function isWireType(declared) {
  const t = (declared || "").trim();
  if (!t) return false;
  const low = t.toLowerCase();

  if (SCALARS.has(low) || BARE_CONTAINERS.has(low)) return true;

  if (t.includes("|") && !t.split("|")[0].includes("[")) {
    return splitTopLevel(t, "|").every(isWireType);
  }

  if (!t.endsWith("]")) return false;
  const open = t.indexOf("[");
  if (open < 0) return false;
  const head = t.slice(0, open).trim().toLowerCase();
  const inner = t.slice(open + 1, -1);

  if (head === "list" || head === "optional") return isWireType(inner);
  if (head === "dict") {
    const parts = splitTopLevel(inner, ",");
    // JSON object keys are always strings — this is the whole reason the check exists.
    return parts.length === 2 && parts[0].toLowerCase() === "str" && isWireType(parts[1]);
  }
  return false;
}

function explainType(declared) {
  const t = (declared || "").trim();
  const low = t.toLowerCase();
  if (low.startsWith("dict[")) {
    return `${t} is not usable across the boundary: JSON object keys are always strings, so a dict must be keyed by str (use Dict[str, ...], or an index-based List[...] instead)`;
  }
  if (low.startsWith("tuple") || low.startsWith("set") || low.startsWith("frozenset")) {
    return `${t} has no JSON equivalent — use List[...] at the boundary and convert inside your code`;
  }
  return `${t} is not a supported wire type. Use int, float, str, bool, None, Any, List[T], Dict[str, T] or Optional[T]`;
}

// Returns a list of human-readable problems; empty means the signature is fine.
// A "class" problem's return_type is the class name and never gets serialized — only
// its constructor args and its methods' signatures actually cross the boundary.
function checkSignature(sig) {
  const problems = [];
  const check = (label, declared) => {
    if (!isWireType(declared)) problems.push(`${label}: ${explainType(declared)}`);
  };

  for (const p of sig.parameters || []) check(`parameter "${p.name}"`, p.type);
  if (sig.kind !== "class") check("return_type", sig.return_type);
  for (const m of sig.methods || []) {
    for (const p of m.parameters || []) check(`method "${m.name}" parameter "${p.name}"`, p.type);
    check(`method "${m.name}" return_type`, m.return_type);
  }
  return problems;
}

export { isWireType, explainType, checkSignature };

File: api/app/wire_types.py
"""The set of declared types that survive the JSON round-trip every value makes.

Test-case inputs travel browser -> /api/verify -> the harness's stdin as JSON, and
results travel back the same way. JSON has no integer-keyed objects, no tuples, no sets
and no classes, so a signature declaring one of those promises the solver something the
pipeline physically cannot deliver — `Dict[int, List[int]]` arrives with string keys and
a solver indexing it by int gets a KeyError.

Before this module the declared type was decorative (rendered in the editor stub, never
checked), and the only guard was a prose instruction to the model. This makes
representability a machine-checked property of the contract instead.

Grammar:
    T := int | float | str | bool | None | Any
       | List[T] | Dict[str, T] | Optional[T] | T | None
"""

_SCALARS = {"int", "float", "str", "bool", "none", "any"}
_BARE_CONTAINERS = {"list", "dict"}


def _split_top_level(text: str) -> list[str]:
    """Split on commas that aren't inside brackets, so Dict[str, List[int]] splits into
    ["str", "List[int]"] rather than on the inner comma."""
    parts, depth, current = [], 0, ""
    for ch in text:
        if ch == "[":
            depth += 1
        elif ch == "]":
            depth -= 1
        if ch == "," and depth == 0:
            parts.append(current)
            current = ""
        else:
            current += ch
    parts.append(current)
    return [p.strip() for p in parts]


def is_wire_type(declared: str) -> bool:
    t = (declared or "").strip()
    if not t:
        return False
    low = t.lower()

    if low in _SCALARS or low in _BARE_CONTAINERS:
        return True

    # `int | None` style unions — every member must itself be representable.
    if "|" in t and "[" not in t.split("|")[0]:
        return all(is_wire_type(part) for part in _split_union(t))

    if not t.endswith("]"):
        return False
    head, _, inner = t.partition("[")
    inner = inner[:-1]
    head = head.strip().lower()

    if head in ("list", "optional"):
        return is_wire_type(inner)
    if head == "dict":
        parts = _split_top_level(inner)
        # The whole point: JSON object keys are strings, so only str keys round-trip.
        return len(parts) == 2 and parts[0].lower() == "str" and is_wire_type(parts[1])
    return False


def _split_union(text: str) -> list[str]:
    parts, depth, current = [], 0, ""
    for ch in text:
        if ch == "[":
            depth += 1
        elif ch == "]":
            depth -= 1
        if ch == "|" and depth == 0:
            parts.append(current)
            current = ""
        else:
            current += ch
    parts.append(current)
    return [p.strip() for p in parts]


def explain(declared: str) -> str:
    """Why a type was rejected, phrased so the model can correct itself — this text is
    fed back into the client's retry loop."""
    t = (declared or "").strip()
    low = t.lower()
    if low.startswith("dict[") and not _split_top_level(t.partition("[")[2][:-1])[0].lower() == "str":
        return (
            f"{t!r} is not usable across the boundary: JSON object keys are always strings, so a "
            "dict must be keyed by str (use Dict[str, ...], or an index-based List[...] instead)"
        )
    if low.startswith("tuple") or low.startswith("set") or low.startswith("frozenset"):
        return f"{t!r} has no JSON equivalent — use List[...] at the boundary and convert inside your code"
    return (
        f"{t!r} is not a supported wire type. Use int, float, str, bool, None, Any, "
        "List[T], Dict[str, T] or Optional[T]"
    )
